libgrust/libproc_macro/registration.h

```diff
@@ -0,0 +1,37 @@
+// Copyright (C) 2023 Free Software Foundation, Inc.
+//
+// This file is part of the GNU Proc Macro Library.  This library is free
+// software; you can redistribute it and/or modify it under the
+// terms of the GNU General Public License as published by the
+// Free Software Foundation; either version 3, or (at your option)
+// any later version.
+
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// Under Section 7 of GPL version 3, you are granted additional
+// permissions described in the GCC Runtime Library Exception, version
+// 3.1, as published by the Free Software Foundation.
+
+// You should have received a copy of the GNU General Public License and
+// a copy of the GCC Runtime Library Exception along with this program;
+// see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
+// <http://www.gnu.org/licenses/>.
+
+#ifndef REGISTRATION_H
+#define REGISTRATION_H
+
+#include <string>
+#include "tokenstream.h"
+
+namespace ProcMacro {
+
+using from_str_function_t = ProcMacro::TokenStream (*) (std::string &, bool &);
```

Comment:
Actually is using std::string and references ok here?

I am probably wrong so please correct me, is this the C API that people use? Because std::string and references wont be available in the C ABI

Reply:
No, we're using the C++ ABI here but I admit it is a little difficult to understand, we're juggling with 3 different ABI at the same time with this library:

- The callback function uses the C++ ABI, We're "communicating" with the C++ libprocmacro internals, not rust, hence why we can use the C++ ABI.
- The symbol containing the address of the callback function uses the C ABI (no mangling)

Does it make sense ? Did I missed a detail ?

Reply:
no thats cool